homeassistant/components/sabnzbd/__init__.py

```diff
@@ -69,42 +106,77 @@ async def async_setup(hass: HomeAssistant, config: ConfigType) -> bool:
     """Set up the SabNzbd Component."""
     sab_api = await get_client(hass, entry.data)
     if not sab_api:
         raise ConfigEntryNotReady
 
+    sab_api_data = SabnzbdApiData(sab_api)
+
     hass.data.setdefault(DOMAIN, {})[entry.entry_id] = {
         KEY_API: sab_api,
+        KEY_API_DATA: sab_api_data,
         KEY_NAME: entry.data[CONF_NAME],
     }
 
-    hass.config_entries.async_setup_platforms(entry, PLATFORMS)
+    @callback
+    def extract_api(func: Callable) -> Callable:
+        """Define a decorator to get the correct api for a service call."""
 
-    sab_api_data = SabnzbdApiData(sab_api)
+        async def wrapper(call: ServiceCall) -> None:
+            """Wrap the service function."""
+            entry_id = async_get_entry_id_for_service_call(hass, call)
+            api_data = hass.data[DOMAIN][entry_id][KEY_API_DATA]
+
+            try:
+                await func(call, api_data)
+            except Exception as err:
+                raise HomeAssistantError(
+                    f"Error while executing {func.__name__}: {err}"
+                ) from err
+
+        return wrapper
+
+    @extract_api
+    async def async_pause_queue(call: ServiceCall, api: SabnzbdApiData) -> None:
+        await api.async_pause_queue()
 
-    async def async_service_handler(service: ServiceCall) -> None:
-        """Handle service calls."""
-        if service.service == SERVICE_PAUSE:
-            await sab_api_data.async_pause_queue()
-        elif service.service == SERVICE_RESUME:
-            await sab_api_data.async_resume_queue()
-        elif service.service == SERVICE_SET_SPEED:
-            speed = service.data.get(ATTR_SPEED)
-            await sab_api_data.async_set_queue_speed(speed)
+    @extract_api
+    async def async_resume_queue(call: ServiceCall, api: SabnzbdApiData) -> None:
+        await api.async_resume_queue()
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_PAUSE, async_service_handler, schema=vol.Schema({})
-    )
+    @extract_api
+    async def async_set_queue_speed(call: ServiceCall, api: SabnzbdApiData) -> None:
+        speed = call.data.get(ATTR_SPEED)
+        await api.async_set_queue_speed(speed)
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_RESUME, async_service_handler, schema=vol.Schema({})
-    )
+    for service, method in (
+        (SERVICE_PAUSE, async_pause_queue),
+        (SERVICE_RESUME, async_resume_queue),
+        (SERVICE_SET_SPEED, async_set_queue_speed),
+    ):
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_SET_SPEED, async_service_handler, schema=SPEED_LIMIT_SCHEMA
-    )
+        if hass.services.has_service(DOMAIN, service):
+            continue
+
+        hass.services.async_register(DOMAIN, service, method, schema=SCHEMA_SERVICE)
+
+        hass.services.async_register(DOMAIN, service, method, schema=SCHEMA_SERVICE)
+
+        hass.services.async_register(DOMAIN, service, method, schema=SCHEMA_SERVICE)
```

Comment:
Is this triple-register correct?

Reply:
We should also have two different service schemas. It's only one service that has the speed limit option.